File: code/ex105.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int i;
    public Map<Integer, Integer> val2idx = new HashMap<>();
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++)
            val2idx.put(inorder[i], i);
        i = 0;
        return build(preorder, 0, inorder.length - 1);
    }
    
    private TreeNode build(int[] preorder, int lo, int hi) {
        if (lo > hi)    return null;
        
        TreeNode root = new TreeNode(preorder[i]);
        int idx = val2idx.get(preorder[i]);
        i++;
        
        root.left = build(preorder, lo, idx - 1);
        root.right = build(preorder, idx + 1, hi);
        
        return root;
    }
}

File: code/ex105.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder, inorder, 0, preorder.size() - 1, 0, inorder.size() - 1);
    }
    
private:
    TreeNode* build(vector<int>& pre, vector<int>& in, int plo, int phi, int ilo, int ihi) {
        if (ilo > ihi)    return NULL;
        
        TreeNode* root = new TreeNode(pre[plo++]);
        
        // Find the pivot to split inorder array to left and right subtrees
        int pivot = 0;
        while (in[pivot] != root->val)    pivot++;
        
        // Build subtrees
        int nleft = pivot - ilo, nright = ihi - pivot - 1;
        root->left = build(pre, in, plo, plo + nleft - 1, ilo, pivot - 1);
        root->right = build(pre, in, plo + nleft, phi, pivot + 1, ihi);
        
        return root;
    }
};